librz/il/definitions/bitvector.c

```diff
@@ -0,0 +1,1078 @@
+		rz_il_bv_free(tmp);
+		return false;
+	}
+
+	for (int i = 0; i < tmp->_elem_len; ++i) {
+		bv->bits[i] = tmp->bits[i];
+	}
+	rz_il_bv_free(tmp);
+
+	return true;
+}
+
+/**
+ * Result of x AND y (`and` operation to every bits)
+ * x and y should have the same length
+ * \param x RzILBitVector, operand
+ * \param y RzILBitVector, operand
+ * \return ret RzILBitVector, a new bitvector, which is the result of AND
+ */
+RZ_API RzILBitVector rz_il_bv_and(RzILBitVector x, RzILBitVector y) {
+	if (x->len != y->len) {
+		return NULL;
+	}
+
+	RzILBitVector ret = rz_il_bv_new(x->len);
+	for (int i = 0; i < ret->_elem_len; ++i) {
+		ret->bits[i] = x->bits[i] & y->bits[i];
+	}
+	return ret;
+}
+
+/**
+ * Result of x OR y (`or` operation to every bits)
+ * x and y should have the same length
+ * \param x RzILBitVector, operand
+ * \param y RzILBitVector, operand
+ * \return ret RzILBitVector, a new bitvector, which is the result of OR
+ */
+RZ_API RzILBitVector rz_il_bv_or(RzILBitVector x, RzILBitVector y) {
+	if (x->len != y->len) {
+		return NULL;
+	}
+
+	RzILBitVector ret = rz_il_bv_new(x->len);
+	for (int i = 0; i < ret->_elem_len; ++i) {
+		ret->bits[i] = x->bits[i] | y->bits[i];
+	}
+	return ret;
+}
+
+/**
+ * Result of x XOR y (`xor` operation to every bits)
+ * x and y should have the same length
+ * \param x RzILBitVector, operand
+ * \param y RzILBitVector, operand
+ * \return ret RzILBitVector, a new bitvector, which is the result of XOR
+ */
+RZ_API RzILBitVector rz_il_bv_xor(RzILBitVector x, RzILBitVector y) {
+	if (x->len != y->len) {
+		return NULL;
+	}
+
+	RzILBitVector ret = rz_il_bv_new(x->len);
+	for (int i = 0; i < ret->_elem_len; ++i) {
+		ret->bits[i] = x->bits[i] ^ y->bits[i];
+	}
+	return ret;
+}
+
+/**
+ * Get the 1's complement of bv
+ * \param bv RzILBitVector, operand
+ * \return ret RzILBitVector, a new bitvector, which is the 1's complement of bv
+ */
+RZ_API RzILBitVector rz_il_bv_complement_1(RzILBitVector bv) {
+	RzILBitVector ret = rz_il_bv_new(bv->len);
+	int real_elem_cnt = bv->_elem_len;
+	for (int i = 0; i < real_elem_cnt; ++i) {
+		ret->bits[i] = ~bv->bits[i];
+	}
+	return ret;
+}
+
+/**
+ * Get the 2's complement of bv
+ * \param bv RzILBitVector, operand
+ * \return ret RzILBitVector, a new bitvector, which is the 2's complement of bv
+ */
+RZ_API RzILBitVector rz_il_bv_complement_2(RzILBitVector bv) {
+	// from right side to left, find the 1st 1 bit
+	// flip/toggle every bit before it
+	RzILBitVector ret = rz_il_bv_dump(bv);
+
+	int i;
+	for (i = bv->len - 1; i > 0; --i) {
+		if (rz_il_bv_get(bv, i) == true) {
+			break;
+		}
+	}
+
+	if (rz_il_bv_get(bv, i) == true) {
+		for (int tmp = 0; tmp < i; ++tmp) {
+			rz_il_bv_toggle(ret, tmp);
+		}
+	}
+
+	return ret;
+}
+
+/**
+ * Result of (x + y) mod 2^length
+ * \param x RzILBitVector, Operand
+ * \param y RzILBitVector, Operand
+ * \return ret RzILBitVector, point to the new bitvector
+ */
+RZ_API RzILBitVector rz_il_bv_add(RzILBitVector x, RzILBitVector y) {
```

Comment:
```suggestion
RZ_API RzILBitVector rz_il_bv_add(RzILBitVector x, RzILBitVector y, bool *cf) {
```
IMO it should be possible to get flag value 
```c
if (cf) {
   *cf = carry;
}

Reply:
probably is better to pass more than just a flag?

Reply:
@wargio explain, please.

By the way, good paper about SMT bitvectors: https://www21.in.tum.de/teaching/sar/SS20/7.pdf

Reply:
i mean, that maybe is faster to standardize the functions to have a specific format., but this is not useful now. so ignore my comment